CODING_STANDARDS.md

```diff
@@ -0,0 +1,139 @@
+// This is necessary because ngAria might add a tabindex to anything with an ng-model
+// (based on whether or not the user has turned that particular feature on/off).
+if (!$attrs['tabindex']) {
+  $element.attr('tabindex', '-1');
+}
+```
+
+#### Prefer more focused, granular components vs. complex, configurable components.
+
+For example, rather than doing this:
+```ts
+<md-button>Basic button</md-button>
+<md-button class="md-fab">FAB</md-button>
+<md-button class="md-icon-button">pony</md-button>
+```
+
+do this:
+```ts
+<md-button>Basic button</md-button>
+<md-fab>FAB</md-fab>
+<md-icon-button>pony</md-icon-button>
+```
+
+#### Prefer small, focused modules
+Keeping modules to a single responsibility makes the code easier to test, consume, and maintain. 
+ES6 modules offer a straightforward way to organize code into logical, granular units. 
+Ideally, individual files are 200 - 300 lines of code.
+
+#### Less is more
+Once a feature is released, it never goes away. We should avoid adding features that don't offer 
+high user value for price we pay both in maintenance, complexity, and payload size. When in doubt, 
+leave it out. 
+
+This applies especially so to providing two different APIs to accomplish the same thing. Always 
+prefer sticking to a _single_ API for accomplishing something. 
+
+### TypeScript
+
+#### Provide function descriptions
+For functions that are more complicated than a simple getter/setter, provide at least a brief 
+sentence explaining what the function does and/or _why_ it does something.
+
+### CSS
+
+#### Be cautious with use of `display: flex`
+* The [baseline calculation for flex elements](http://www.w3.org/TR/css-flexbox-1/#flex-baselines) 
+is different than other display values, making it difficult to align flex elements with standard 
+elements like input and button.
+* Component outermost elements are never flex (block or inline-block)
+* Don't use `display: flex` on elements that will contain projected content.
+
+#### Use lowest specificity possible
+Always prioritize lower specificity over other factors. Most style definitions should consist of a 
+single element or css selector plus necessary state modifiers. Avoid SCSS nesting for the sake of 
+code organization. This will allow users to much more easily override styles.
+
+For example, rather than doing this:
+```scss
+md-calendar {
+  display: block;
+
+  .md-month {
+    display: inline-block;
+
+    .md-date.md-selected {
+      font-weight: bold;
+    }
+  }
+}
+```
+
+do this:
+```scss
+md-calendar {
+  display: block;
+}
+
+.md-calendar-month {
+  display: inline-block;
+}
+
+.md-calendar-date.md-selected {
+  font-weight: bold;
+}
+```
+
+#### Never set a margin on a host element.
+The end-user of a component should be the one to decide how much margin a component has around it.
+
+#### Apply styles to the host element when it is equivalent to apply them within the template.
```

Comment:
Can we make this wording a bit clearer?  Maybe `Apply styles to the host element rather than its children when possible.` with more explanation in the body?